src/index.js

```diff
@@ -3,247 +3,267 @@
-              input = asset.source();
-              uglify.AST_Node.warn_function = (warning) => { // eslint-disable-line camelcase
-                warnings.push(warning);
-              };
-            }
-            uglify.base54.reset();
-            let ast = uglify.parse(input, {
-              filename: file,
-            });
-            if (options.compress !== false) {
-              ast.figure_out_scope();
-              const compress = uglify.Compressor(options.compress || {
-                warnings: false,
-              }); // eslint-disable-line new-cap
-              ast = compress.compress(ast);
-            }
-            if (options.mangle !== false) {
-              ast.figure_out_scope(options.mangle || {});
-              ast.compute_char_frequency(options.mangle || {});
-              ast.mangle_names(options.mangle || {});
-              if (options.mangle && options.mangle.props) {
-                uglify.mangle_properties(ast, options.mangle.props);
+
+              // Handling comment extraction
+              const extractedComments = [];
+              let commentsFile = false;
+              if (this.options.extractComments) {
+                uglifyOptions.output = uglifyOptions.output || {};
+                uglifyOptions.output.comments = UglifyJsPlugin.buildCommentsFunction(this.options, uglifyOptions, extractedComments);
+
+                commentsFile = this.options.extractComments.filename || `${file}.LICENSE`;
+                if (typeof commentsFile === 'function') {
+                  commentsFile = commentsFile(file);
+                }
               }
-            }
-            const output = {};
-            output.comments = Object.prototype.hasOwnProperty.call(options, 'comments') ? options.comments : /^\**!|@preserve|@license/;
-            output.beautify = options.beautify;
-            for (const k in options.output) { // eslint-disable-line guard-for-in
-              output[k] = options.output[k];
-            }
-            const extractedComments = [];
-            if (options.extractComments) {
-              const condition = {};
-              if (typeof options.extractComments === 'string' || options.extractComments instanceof RegExp) {
-                // extractComments specifies the extract condition and output.comments specifies the preserve condition
-                condition.preserve = output.comments;
-                condition.extract = options.extractComments;
-              } else if (Object.prototype.hasOwnProperty.call(options.extractComments, 'condition')) {
-                // Extract condition is given in extractComments.condition
-                condition.preserve = output.comments;
-                condition.extract = options.extractComments.condition;
+
+              // Calling uglify
+              const { error, map, code, warnings } = uglify.minify({ [file]: input }, uglifyOptions);
+
+              // Handling results
+              if (error) {
```

Comment:
instead of `throw` and immediate `catch`, perhaps you can do:

```
if (error) {
  compilation.errors.push(UglifyJsPlugin.buildError(error, file, sourceMap, compilation, requestShortener));
  return;
}
```

here. This will avoid a try/catch deoptimization until everyone is on Node 8 with V8 that has TF+I that can properly optimize try/catch statements.